specs/dwritecore/DWriteCore.md

```diff
@@ -1050,9 +1258,604 @@ int __cdecl wmain(int argc, WCHAR** argv)
 }
 ```
 
-# API Details
+### Rendering Color Glyph Runs
+
+This section shows a sample `TextRenderer::DrawGlyphRun` method that implements the `DrawGlyphRun`
+method of the `IDWriteTextRenderer` interface. The `IDWriteTextRenderer` callback interface is used
+by `IDWriteTextLayout::Draw` to render glyph runs, underlines, and so on. The use of a callback
+interface enables the text layout API to be decoupled from the specific graphics engine. For
+example, it is possible to implement `IDWriteTextRenderer` in terms of Direct2D, in terms of
+`IDWriteBitmapRenderTarget`, or in some other way. This example uses Direct2D.
+
+The example includes only the `DrawGlyphRun` method. The rest of the `TextRenderer` class is omitted
+for brevity, but the example assumes the class implements the `IDWriteTextRenderer` interface and
+has the following member variables:
+
+```c++
+wil::com_ptr<IDWriteFactory4> m_dwriteFactory;
+wil::com_ptr<ID2D1DeviceContext4> m_deviceContext;
+wil::com_ptr<ID2D1Brush> m_foregroundBrush;
+```
+
+The example method renders color glyphs in one of two ways. The easy way is to call
+`ID2D1DeviceContext7::DrawGlyphRunWithColor`. However, the example also includes a fallback code
+path in case the available version of Direct2D does not implement the `ID2D1DeviceContext7`
+interface. The fallback code path uses `TranslateColorGlyphRun`.
+
+```c++
+// Implementation of IDWriteTextRenderer::DrawGlyphRun
+HRESULT STDMETHODCALLTYPE TextRenderer::DrawGlyphRun(
+    _In_opt_ void* clientDrawingContext,
+    FLOAT baselineOriginX,
+    FLOAT baselineOriginY,
+    DWRITE_MEASURING_MODE measuringMode,
+    _In_ DWRITE_GLYPH_RUN const* glyphRun,
+    _In_ DWRITE_GLYPH_RUN_DESCRIPTION const* glyphRunDescription,
+    _In_opt_ IUnknown* clientDrawingEffect
+) noexcept
+{
+    D2D_POINT_2F baselineOrigin{ baselineOriginX, baselineOriginY };
+
+    // If a brush is specified via the clientDrawingEffect parameter, use it as
+    // the foreground brush for this glyph run. Otherwise, use the brush specified
+    // by the m_foregroundBrush member.
+    wil::com_ptr<ID2D1Brush> foregroundBrush;
+    if (clientDrawingEffect != nullptr)
+    {
+        foregroundBrush = wil::try_com_query<ID2D1Brush>(clientDrawingEffect);
+    }
+    if (foregroundBrush == nullptr)
+    {
+        foregroundBrush = m_foregroundBrush;
+    }
+
+    // If available, use the ID2D1DeviceContext7 interface to draw
+    // the glyph run with color the "easy" way.
+    if (auto deviceContext7 = m_deviceContext.try_query<ID2D1DeviceContext7>())
+    {
+        deviceContext7->DrawGlyphRunWithColor(
```

Comment:
SAMPLE: Yes

Reply:
The sample is actually correct because D2D drawing methods return void. If a drawing operation fails, an error is returned from EndDraw. I'm adding a comment to that effect. (I think this design is because drawing operations are batched, so a failure may not be known until the device context is flushed.)